unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,699 @@
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder. Implements `Default` which tracks the recent defaults.
+///
+/// Custom file tree builder can be created with [`FileAdder::builder()`] and configuring the
+/// chunker and collector.
+///
+/// Current implementation maintains an internal buffer for the block creation and uses a
+/// non-customizable hash function to produce Cid version 0 links. Currently does not support
+/// inline links.
+#[derive(Default)]
+pub struct FileAdder {
+    chunker: Chunker,
+    collector: Collector,
+    block_buffer: Vec<u8>,
+    // all unflushed links as a flat vec; this is compacted as we grow and need to create a link
+    // block for the last N blocks, as decided by the collector
+    unflushed_links: Vec<(usize, Cid, u64, u64)>,
+}
+
+impl fmt::Debug for FileAdder {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "FileAdder {{ chunker: {:?}, block_buffer: {}/{}, unflushed_links: {} }}",
+            self.chunker,
+            self.block_buffer.len(),
+            self.block_buffer.capacity(),
+            LinkFormatter(&self.unflushed_links),
+        )
+    }
+}
+
+struct LinkFormatter<'a>(&'a [(usize, Cid, u64, u64)]);
+
+impl fmt::Display for LinkFormatter<'_> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut iter = self.0.iter().peekable();
+
+        write!(fmt, "[")?;
+
+        let mut current = match iter.peek() {
+            Some((depth, ..)) => depth,
+            None => return write!(fmt, "]"),
+        };
+
+        let mut count = 0;
+
+        for (next_depth, ..) in iter {
+            if current == next_depth {
+                count += 1;
+            } else {
+                write!(fmt, "{}/", count)?;
+                for _ in *next_depth..(*current - 1) {
+                    write!(fmt, "0/")?;
+                }
+                count = 1;
+                current = next_depth;
+            }
+        }
+
+        write!(fmt, "{}]", count)
+    }
+}
+
+/// Convinience type to facilitate configuring FileAdders
+#[derive(Default)]
+pub struct FileAdderBuilder {
+    chunker: Chunker,
+    collector: Collector,
+}
+
+impl FileAdderBuilder {
+    /// Configures the builder to use the given chunker.
+    pub fn set_chunker(&mut self, chunker: Chunker) -> &mut Self {
+        self.chunker = chunker;
+        self
+    }
+
+    /// Configures the builder to use the given collector or layout.
+    pub fn set_collector(&mut self, collector: Collector) -> &mut Self {
+        self.collector = collector;
+        self
+    }
+
+    /// Returns a new FileAdder
+    pub fn build(&self) -> FileAdder {
+        let chunker = self.chunker.clone();
+        let collector = self.collector.clone();
+
+        FileAdder {
+            chunker,
+            collector,
+            block_buffer: Vec::new(),
```

Comment:
the last two objects can probably be initialized with `..Default::default()` here, i.e.:
```
FileAdder {
            chunker,
            collector,
            ..Default::default(),
}
```

Reply:
Yes, thanks. Was going back and forth with this one, does the block_buffer need to be created with an allocation or not.

This change cannot be done for the BalancedCollector::with_branching_factor without stackoverflow.

Reply:
Yeah, it's because the `Default` impl calls `with_branching_factor`, so you have `with_branching_factor` calling `with_branching_factor` calling `with_branching_factor` etc.

Reply:
I know what you mean, but the compiler isn't smart enough yet to see this doesn't have to recurse because `branching_factor` is already known

Reply:
Aah good point. Didn't think the compiler would actually "call" std::default::Default when using `..Default::default()` but would instead pick up the expressions to produce the not mentioned fields... Well, I guess that expression is the one which yielded the stackoverflow. Perhaps I would had expected a warning on unbounded recursion then as this is a compiler implementation detail how the fields are handled. (I certainly don't expect compiler to create a new value at runtime and then to merge some fields ... That'd be quite wasteful and make me not want to use this syntax.)